library/std/src/io/mod.rs

```diff
@@ -2924,6 +2924,22 @@ impl<T: Read> Read for Take<T> {
         Ok(n)
     }
 
+    fn read_exact(&mut self, buf: &mut [u8]) -> Result<()> {
+        // Don't call into inner reader at all at EOF because it may still block
+        if self.limit == 0 {
+            return Ok(());
+        }
+
+        let len = buf.len() as u64;
+        if self.limit > len {
```

Comment:
```suggestion
        if self.limit >= len {
```
Otherwise, this will return `Err(EOF)` when the limit is exactly enough.